databricks/koalas/tests/test_indexes.py

```diff
@@ -1390,6 +1390,118 @@ def test_hasnans(self):
         kser = ks.from_pandas(pser)
         self.assert_eq(pser.hasnans, kser.hasnans)
 
+    def test_intersection(self):
+        pidx = pd.Index([1, 2, 3, 4], name="Koalas")
+        kidx = ks.from_pandas(pidx)
+
+        # other = Index
+        pidx_other = pd.Index([3, 4, 5, 6], name="Koalas")
+        kidx_other = ks.from_pandas(pidx_other)
+        self.assert_eq(pidx.intersection(pidx_other), kidx.intersection(kidx_other).sort_values())
+        self.assert_eq(
+            (pidx + 1).intersection(pidx_other), (kidx + 1).intersection(kidx_other).sort_values()
+        )
+
+        pidx_other_different_name = pd.Index([3, 4, 5, 6], name="Databricks")
+        kidx_other_different_name = ks.from_pandas(pidx_other_different_name)
+        self.assert_eq(
+            pidx.intersection(pidx_other_different_name),
+            kidx.intersection(kidx_other_different_name).sort_values(),
+        )
+        self.assert_eq(
+            (pidx + 1).intersection(pidx_other_different_name),
+            (kidx + 1).intersection(kidx_other_different_name).sort_values(),
+        )
+
+        pidx_other_from_frame = pd.DataFrame({"a": [3, 4, 5, 6]}).set_index("a").index
+        kidx_other_from_frame = ks.from_pandas(pidx_other_from_frame)
+        self.assert_eq(
+            pidx.intersection(pidx_other_from_frame),
+            kidx.intersection(kidx_other_from_frame).sort_values(),
+        )
+        self.assert_eq(
+            (pidx + 1).intersection(pidx_other_from_frame),
+            (kidx + 1).intersection(kidx_other_from_frame).sort_values(),
+        )
+
+        # other = MultiIndex
+        pmidx = pd.MultiIndex.from_tuples([("a", "x"), ("b", "y"), ("c", "z")])
+        kmidx = ks.from_pandas(pmidx)
+        self.assert_eq(
+            pidx.intersection(pmidx), kidx.intersection(kmidx).sort_values(), almost=True
```

Comment:
Why `almost=True`?

Reply:
Otherwise the tests are failure even they are same Empty `MultiIndex` with `AssertionError` below.

```
AssertionError: MultiIndex level [0] are different

MultiIndex level [0] classes are not equivalent
[left]:  Index([], dtype='object')
[right]: Int64Index([], dtype='int64')

Left:
MultiIndex([], )
object

Right:
MultiIndex([], )
object
```

Reply:
After several tries, I'd say maybe `assert_eq` cannot compare the empty `MultiIndex` properly for some reason.

They don't even show the reason for `AssertionError` show as below  😅

```
    raise AssertionError(msg) from e
AssertionError: 

Left:
MultiIndex([], )
object

Right:
MultiIndex([], )
object
```

Let me investigate this more in the separated PR